### DISCUSSION TITLE
thoughts about this macro?

### DISCUSSION THREAD

**Discussion by stylewarning** [coalton-lang/coalton#552] [Ideas and Proposals]
Any thoughts (aside from its terrible name) about this macro which defines a monomorphization and specialization at the same time?

```
(cl:defmacro spec (f cl:&rest types)
  `(coalton-toplevel
     ,@(cl:loop
          :for type :in types
          :for g    := (cl:gentemp (cl:symbol-name f))
          :collect `(monomorphize)
          :collect `(declare ,g ,type)
          :collect `(define ,g ,f)
          :collect `(specialize ,f ,g ,type))))
```

For example:

```
(define (f x)
  (+ 1 (* 2 x)))

(spec f
      (Double-Float -> Double-Float)
      (Single-Float -> Single-Float))
```

**Comment by eliaslfox** [comment]
I would prefer to hold off on defining something like this. Application use of Coalton should really be monomorphizing once at the programs entry point. In part because the `monomorphize` attribute seems like a good place for other whole program optimizations, such as recompiling functions with different aritys to avoid currying or closures. See [this post](https://blog.janestreet.com/flambda/) for more possibilities. Monomorphization also breaks function redefinition, so calling the monomorphizer multiple times in a single projects call tree could lead to unexpected results.

This does seem potentially useful for library writers, but I think we should wait until a library ecosystem develops before preemptively adding macros to support it.
